runtime/server/psql.go

```diff
@@ -0,0 +1,252 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v5/pgtype"
+	wire "github.com/jeroenrinzema/psql-wire"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/server/auth"
+	"github.com/rilldata/rill/runtime/server/psql"
+)
+
+// psqlQueryHandler handles incoming queries on psql wire endpoint.
+// The metadata queries are redirected to ResolvePSQLQuery.
+// The metrics_sql queries are redirected to metrics_sql resolver.
+func (s *Server) psqlQueryHandler(ctx context.Context, query string) (wire.PreparedStatements, error) {
+	clientParams := wire.ClientParameters(ctx)
+	instanceID := clientParams[wire.ParamDatabase]
+
+	// The logic to identify metadata queries is somewhat hacky.
+	// Identify metadata queries based on whether query contains names of common postgres catalogs like pg_attribute, pg_catalog etc.
+	// Also consider queries that do not have a FROM clause as metadata queries since querying from metrics_view requires a FROM clause.
+	if strings.Contains(query, "pg_attribute") || strings.Contains(query, "pg_catalog") || strings.Contains(query, "pg_type") || strings.Contains(query, "pg_namespace") || !strings.Contains(strings.ToUpper(query), "FROM") {
+		data, schema, err := psql.ResolvePSQLQuery(ctx, &psql.PSQLQueryOpts{
+			SQL:            query,
+			Runtime:        s.runtime,
+			InstanceID:     instanceID,
+			UserAttributes: auth.GetClaims(ctx).Attributes(),
+			Priority:       1,
+			Logger:         s.logger,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		handle := func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			for i := 0; i < len(data); i++ {
+				if err := writer.Row(data[i]); err != nil {
+					return fmt.Errorf("failed to write row: %w", err)
+				}
+			}
+			return writer.Complete("OK")
+		}
+		return wire.Prepared(wire.NewStatement(handle, wire.WithColumns(convertSchema(schema)))), nil
+	}
+
+	// todo handle normal SQL. Some ideas to handle that
+	// 1. Approach similar to BigQuery which has #legacySQL
+	// 2. We can parse for a -- @connector: xxx comment, and if found, use that connector directly instead of metrics SQL
+	res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+		InstanceID:         instanceID,
+		Resolver:           "metrics_sql",
+		ResolverProperties: map[string]any{"sql": query},
+		Args:               map[string]any{"priority": 1},
+		UserAttributes:     auth.GetClaims(ctx).Attributes(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return wire.Prepared(wire.NewStatement(s.execQuery(instanceID, query), wire.WithColumns(convertSchema(res.Schema)))), nil
+}
+
+func (s *Server) execQuery(instanceID, query string) wire.PreparedStatementFn {
```

Comment:
Can we name this something like `psqlExecQuery`? The current name would be fine if this was a dedicated `psql` package, but when it's in `runtime/server`, which is one of our larger packages, it's lacking context. The usual rule of thumb is name length should be inversely proportional to the scope. The scope of `*Server` is some 20 files, so worth adding more details here.

Reply:
Sure. I think in that case will add these as methods of a struct.